sfe/sfe.go

```diff
@@ -0,0 +1,287 @@
+package sfe
+
+import (
+	"crypto"
+	"crypto/ed25519"
+	"embed"
+	"errors"
+	"fmt"
+	"io/fs"
+	"net/http"
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/go-jose/go-jose/v4"
+	"github.com/go-jose/go-jose/v4/jwt"
+	"github.com/jmhodges/clock"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+
+	"github.com/letsencrypt/boulder/core"
+
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/metrics/measured_http"
+	rapb "github.com/letsencrypt/boulder/ra/proto"
+	sapb "github.com/letsencrypt/boulder/sa/proto"
+)
+
+const (
+	// The API version should be checked when parsing parameters to quickly deny
+	// a client request. Can be used to mass-invalidate URLs.
+	unpausePath = "/sfe/v1/unpause"
+)
+
+var (
+	//go:embed all:static
+	staticFS embed.FS
+
+	//go:embed all:templates all:pages
+	dynamicFS embed.FS
+
+	// HTML pages to-be-served by the SFE
+	tmplPages *template.Template
+)
+
+// Parse the files once at startup to avoid each request causing the server to
+// JIT parse. The pages are stored in an in-memory embed.FS to prevent
+// unnecessary filesystem I/O on a physical HDD.
+func init() {
+	tmplPages = template.Must(template.New("pages").ParseFS(dynamicFS, "templates/layout.html", "pages/*"))
+}
+
+// SelfServiceFrontEndImpl provides all the logic for Boulder's selfservice
+// frontend web-facing interface, i.e., a portal where a subscriber can unpause
+// their account. Its methods are primarily handlers for HTTPS requests for the
+// various non-ACME functions.
+type SelfServiceFrontEndImpl struct {
+	ra rapb.RegistrationAuthorityClient
+	sa sapb.StorageAuthorityReadOnlyClient
+
+	log blog.Logger
+	clk clock.Clock
+
+	// requestTimeout is the per-request overall timeout.
+	requestTimeout time.Duration
+
+	// CORS settings
+	AllowOrigins []string
+
+	// unpausePubKey is a x/crypto/ed25519 public key derived from a seed value
+	// shared by the SFE and WFE. It is used to validate incoming JWT signatures
+	// on the unpause endpoint.
+	unpausePubKey crypto.PublicKey
+}
+
+// NewSelfServiceFrontEndImpl constructs a web service for Boulder
+func NewSelfServiceFrontEndImpl(
+	stats prometheus.Registerer,
+	clk clock.Clock,
+	logger blog.Logger,
+	requestTimeout time.Duration,
+	rac rapb.RegistrationAuthorityClient,
+	sac sapb.StorageAuthorityReadOnlyClient,
+	unpauseSeed string,
+) (SelfServiceFrontEndImpl, error) {
+	// The seed is used to generate an x/crypto/ed25519 keypair which
+	// requires a SeedSize of 32 bytes or the generator will panic.
+	if len(unpauseSeed) != 32 {
+		return SelfServiceFrontEndImpl{}, errors.New("unpauseSeed should be 32 hexadecimal characters e.g. the output of 'openssl rand -hex 16'")
+	}
+
+	// We only need the public key to check the JWT signature and it only needs
+	// to be generate at startup.
+	unpausePubKey := ed25519.NewKeyFromSeed([]byte(unpauseSeed)).Public()
```

Comment:
Following up on my previous comment, do all of this work inside of a method bound to the config struct that acts as an accessor.

```
type Unpause struct {
	// Seed is a secret that should contain 256 bits (32 bytes) of random data
	// used to derive an x/crypto/ed25519 keypair (e.g. the output of `openssl
	// rand -hex 16`). In a multi-DC deployment this value should be the same
	// across all boulder-wfe and sfe instances.
	Seed PasswordConfig `validate:"-"`
}

func (u *Unpause) UnpausePubKey() (crypto.PublicKey, error) {
	seed, err := u.Seed.Pass()
	if err != nil {
		return nil, err
	}
	if len(seed) != 32 {
		return nil, errors.New("unpause seed should be 32 hexadecimal characters e.g. the output of 'openssl rand -hex 16'")
	}
	return ed25519.NewKeyFromSeed([]byte(seed)).Public(), nil
}
```